### PR TITLE
Flp 489 reject start date request employer

### PR DESCRIPTION
**Pull request by tom-gough**

### CODE REVIEW

File: src/Command/RejectStartDateChange/RejectStartDateChangeCommand.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+﻿namespace SFA.DAS.Apprenticeships.Command.RejectStartDateChange
+{
```
Comment (MikeV-TalentConsulting) [inline src/Command/RejectStartDateChange/RejectStartDateChangeCommand.cs:2]: Could use File scoped namespace here

File: src/Domain/Apprenticeship/Events/PriceChangeApprovedHandler.cs [hunk 2]
```diff
@@ -37,4 +37,37 @@ public async Task Handle(PriceChangeApproved @event, CancellationToken cancellat
             await _messageSession.Publish(apprenticeshipCreatedEvent);
         }
     }
+
+    public class StartDateChangeApprovedHandler : IDomainEventHandler<StartDateChangeApproved>
```
Comment (MikeV-TalentConsulting) [inline src/Domain/Apprenticeship/Events/PriceChangeApprovedHandler.cs:41]: Any reason this isn't in its own file? or in the same file as the event?

File: src/Domain/Apprenticeship/Events/PriceChangeApprovedHandler.cs [hunk 3]
```diff
@@ -37,4 +37,37 @@ public async Task Handle(PriceChangeApproved @event, CancellationToken cancellat
...
+            {
+                ApprenticeshipKey = apprenticeship.Key,
+                ApprenticeshipId = approval.ApprovalsApprenticeshipId,
+                EmployerAccountId = apprenticeship.EmployerAccountId,
+                ApprovedDate = @event.ApprovedBy == ApprovedBy.Employer ? startDateChange.EmployerApprovedDate!.Value : startDateChange.ProviderApprovedDate!.Value,
+                ProviderId = apprenticeship.Ukprn,
+                ActualStartDate = startDateChange.ActualStartDate,
+                ProviderApprovedBy = startDateChange.ProviderApprovedBy,
+                EmployerApprovedBy = startDateChange.EmployerApprovedBy,
+                Initiator = startDateChange.Initiator.ToString()
```
Comment (MikeV-TalentConsulting) [inline src/Domain/Apprenticeship/Events/PriceChangeApprovedHandler.cs:67]: Code warning

File: src/Domain/Apprenticeship/StartDateChangeDomainModel.cs [hunk 4]
```diff
@@ -1,62 +1,84 @@
...
+public class StartDateChangeDomainModel
 {
-    public class StartDateChangeDomainModel
+    private readonly StartDateChange _entity;
+
+    public Guid Key => _entity.Key;
+    public Guid ApprenticeshipKey => _entity.ApprenticeshipKey;
+    public DateTime ActualStartDate => _entity.ActualStartDate;
+    public string Reason => _entity.Reason;
+    public string ProviderApprovedBy => _entity.ProviderApprovedBy;
```
Comment (MikeV-TalentConsulting) [inline src/Domain/Apprenticeship/StartDateChangeDomainModel.cs:14]: Code warning

File: src/Domain/Apprenticeship/StartDateChangeDomainModel.cs [hunk 5]
```diff
@@ -1,62 +1,84 @@
...
+    public DateTime? ProviderApprovedDate => _entity.ProviderApprovedDate;
+    public string EmployerApprovedBy => _entity.EmployerApprovedBy;
```
Comment (MikeV-TalentConsulting) [inline src/Domain/Apprenticeship/StartDateChangeDomainModel.cs:16]: Code warning

File: src/Domain/Apprenticeship/StartDateChangeDomainModel.cs [hunk 6]
```diff
@@ -1,62 +1,84 @@
...
+            _entity.ProviderApprovedBy = userId;
+            _entity.ProviderApprovedDate = approvedDate;
         }
     }
-}
+
+    public void Reject(string? reason)
+    {
+        _entity.RequestStatus = ChangeRequestStatus.Rejected;
+        _entity.RejectReason = reason;
```
Comment (MikeV-TalentConsulting) [inline src/Domain/Apprenticeship/StartDateChangeDomainModel.cs:82]: Code warning

File: src/Domain/Repositories/ApprenticeshipRepository.cs [hunk 7]
```diff
@@ -39,6 +39,7 @@ public async Task<ApprenticeshipDomainModel> Get(Guid key)
         var apprenticeship = await DbContext.Apprenticeships
             .Include(x => x.Approvals)
             .Include(x => x.PriceHistories)
+                .Include(x => x.StartDateChanges)
```
Comment (MikeV-TalentConsulting) [inline src/Domain/Repositories/ApprenticeshipRepository.cs:42]: minor, but maybe worth matching the indentation

File: src/Command/RejectStartDateChange/RejectStartDateChangeCommand.cs (hunk 1 above)
Comment (tom-gough) [reply src/Command/RejectStartDateChange/RejectStartDateChangeCommand.cs:2]: sorted

File: src/Domain/Apprenticeship/Events/PriceChangeApprovedHandler.cs [hunk 8]
```diff
@@ -37,4 +37,37 @@ public async Task Handle(PriceChangeApproved @event, CancellationToken cancellat
             await _messageSession.Publish(apprenticeshipCreatedEvent);
         }
     }
+
```
[the remark this answers is not in the document]
Comment (tom-gough) [reply src/Domain/Apprenticeship/Events/PriceChangeApprovedHandler.cs:40]: sorted

File: src/Domain/Apprenticeship/Events/PriceChangeApprovedHandler.cs (hunk 2 above)
Comment (tom-gough) [reply src/Domain/Apprenticeship/Events/PriceChangeApprovedHandler.cs:41]: sorted

File: src/Domain/Apprenticeship/Events/PriceChangeApprovedHandler.cs (hunk 3 above)
Comment (tom-gough) [reply src/Domain/Apprenticeship/Events/PriceChangeApprovedHandler.cs:67]: sorted

File: src/Domain/Apprenticeship/StartDateChangeDomainModel.cs (hunk 4 above)
Comment (tom-gough) [reply src/Domain/Apprenticeship/StartDateChangeDomainModel.cs:14]: sorted

File: src/Domain/Apprenticeship/StartDateChangeDomainModel.cs (hunk 5 above)
Comment (tom-gough) [reply src/Domain/Apprenticeship/StartDateChangeDomainModel.cs:16]: sorted

File: src/Domain/Apprenticeship/StartDateChangeDomainModel.cs (hunk 6 above)
Comment (tom-gough) [reply src/Domain/Apprenticeship/StartDateChangeDomainModel.cs:82]: sorted

**Review by MikeV-TalentConsulting — APPROVED** [review]

**Pull request merged**